{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We initially used a common 2-state markov model and built out viterbi and Baum-Welch estimation following the general outline in the book. This model, however, proved largely inaccurate as it only measures GC-content of a sequence. The current model has eight different states for each combination of possible emits and CpG regions. This allows it to measure dinucleotide frequency. The transition matrix functionally contains the probabilities of emits for any given state as well as CpG and non-CpG changeover. The emission matrix functionally has no real probabilities involved. It is difficult to find information on look-back Markov Models in general, the information used to originally implement this was found here:\n",
    "https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-047-computational-biology-fall-2015/lectures_slides/MIT6_047F15_Lecture05.pdf\n",
    "Baum-welch (or any EM learning) proved too difficult to implement in look-back models for the time being.\n",
    "\n",
    "The algorithm is being run on the region of chromosome 6 containing the TNXB referenced in our commonly methylated genes list.\n",
    "Runtime on ~68,000 base sequence:\n",
    "        real    0m0.770s\n",
    "        user    0m0.750s\n",
    "        sys     0m0.031s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: bad file or filepath\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import math\n",
    "\n",
    "states = ['A','C','G','T','a','c','g','t']\n",
    "emits = {'A':0,'C':1,'G':2,'T':3}\n",
    "\n",
    "def viterbi(nucs,tMat, eMat, initProbs):\n",
    "    nucCnt = len(nucs)\n",
    "    stateCnt = len(states)\n",
    "    trellis = [[0.0 for i in range(stateCnt)] for j in range(nucCnt)]\n",
    "    backtrack = [[0.0 for i in range(stateCnt)] for j in range(nucCnt)]\n",
    "    tMat = [[math.log(val) for val in row] for row in tMat]\n",
    "    eMat = [[math.log(val) for val in row] for row in eMat]\n",
    "    for i in range(stateCnt):\n",
    "        trellis[0][i] = eMat[i][emits[nucs[0]]] + math.log(initProbs[i]) #weights for initial state\n",
    "\n",
    "    #fill trellis\n",
    "    for i in range(1,nucCnt):\n",
    "        for j in range(stateCnt):\n",
    "            heavy = -math.inf\n",
    "            heavy_idx = 0\n",
    "            for k in range(stateCnt):\n",
    "                temp = trellis[i-1][k] + tMat[k][j] #accessing\n",
    "                if temp > heavy:\n",
    "                    heavy = temp\n",
    "                    heavy_idx = k\n",
    "            trellis[i][j] = heavy + eMat[j][emits[nucs[i]]] #setting\n",
    "            backtrack[i][j] = heavy_idx\n",
    "\n",
    "    #walk back through heaviest trellis\n",
    "    idxPrediction = [0]*nucCnt\n",
    "    heaviest = -math.inf\n",
    "    heaviest_idx = 0\n",
    "    for i in range(stateCnt):\n",
    "        if trellis[nucCnt-1][i] > heaviest:\n",
    "            heaviest = trellis[nucCnt - 1][i]\n",
    "            heaviest_idx = i\n",
    "    idxPrediction[nucCnt-1] = heaviest_idx\n",
    "    for i in range(nucCnt-1,0,-1):\n",
    "        idxPrediction[i-1] = backtrack[i][idxPrediction[i]]\n",
    "    statePrediction = ''.join(states[i] for i in idxPrediction)\n",
    "    return statePrediction\n",
    "\n",
    "\n",
    "def main(argv):\n",
    "    #not particularly important for percentages to sum correctly. Relative to weighting\n",
    "    #only allowed to go +->- in G state (end of dinucleotide) and -->+ in c state (start of dinucleotide)\n",
    "            #A         C       G        T        a        c         g        t\n",
    "    tMat_default =[\n",
    "         [0.18000, 0.27400, 0.42600, 0.12000, 1.0e-37, 1.0e-37, 1.0e-37, 1.0e-37], #A\n",
    "         [0.17100, 0.36800, 0.27400, 0.18800, 1.0e-37, 1.0e-37, 1.0e-37, 1.0e-37], #C\n",
    "         [0.16100, 0.33900, 0.37500, 0.12500, 0.00100, 0.00100, 0.00100, 0.00100], #G\n",
    "         [0.07900, 0.35500, 0.38400, 0.18200, 1.0e-37, 1.0e-37, 1.0e-37, 1.0e-37], #T\n",
    "         [1.0e-37, 0.00025, 1.0e-37, 1.0e-37, 0.30000, 0.20500, 0.28500, 0.21000], #a\n",
    "         [1.0e-37, 0.00025, 1.0e-37, 1.0e-37, 0.32200, 0.29800, 0.07800, 0.30200], #c\n",
    "         [1.0e-37, 0.00025, 1.0e-37, 1.0e-37, 0.24800, 0.24600, 0.29800, 0.20800], #g\n",
    "         [1.0e-37, 0.00025, 1.0e-37, 1.0e-37, 0.17700, 0.23900, 0.29200, 0.29200]  #t\n",
    "        ]\n",
    "    #transition matrix values for 8-state look-back is very hard to find information on. Above are placeholders that perform relatively well but need adjusting.\n",
    "    #difficult to perform traditional EM to generate as next emission probabilities are abstracted into transition matrix.\n",
    "    #intra-state transition weights taken\n",
    "    #from https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-047-computational-biology-fall-2015/lectures_slides/MIT6_047F15_Lecture05.pdf\n",
    "    #cross-state weights estimated from textbook information and could use some tuning\n",
    "\n",
    "\n",
    "    #emission matrix is sort of silly for 8 state look-back, but keeping it for clarity\n",
    "    #0s->lower end of 4 bytes (1e-37-ish) for log-space\n",
    "    eMat_default = [\n",
    "          [1.00000, 1.0e-37, 1.0e-37, 1.0e-37],\n",
    "          [1.0e-37, 1.00000, 1.0e-37, 1.0e-37],\n",
    "          [1.0e-37, 1.0e-37, 1.00000, 1.0e-37],\n",
    "          [1.0e-37, 1.0e-37, 1.0e-37, 1.00000],\n",
    "          [1.00000, 1.0e-37, 1.0e-37, 1.0e-37],\n",
    "          [1.0e-37, 1.00000, 1.0e-37, 1.0e-37],\n",
    "          [1.0e-37, 1.0e-37, 1.00000, 1.0e-37],\n",
    "          [1.0e-37, 1.0e-37, 1.0e-37, 1.00000]\n",
    "        ]\n",
    "\n",
    "    #placeholders again. Perform well enough. Estimated from book.\n",
    "    #                       A        C        G        T        a        c        g        t\n",
    "    initProbs_default = [0.01500, 0.03000, 0.03000, 0.01500, 0.25000, 0.21000, 0.21000, 0.25000]\n",
    "\n",
    "    try:\n",
    "        with open(argv[0],'r') as f:\n",
    "            gName = f.readline().strip()\n",
    "            nucs = \"\".join(f.read().splitlines())\n",
    "    except:\n",
    "        print(\"usage: bad file or filepath\")\n",
    "        return\n",
    "    statePrediction = viterbi(nucs,tMat_default,eMat_default,initProbs_default)\n",
    "\n",
    "    #finds beginning and end of islands from state list\n",
    "    print(gName)\n",
    "    if statePrediction[0].isupper():\n",
    "        island_start = 0\n",
    "    else:\n",
    "        island_start = None\n",
    "    for i in range(1, len(statePrediction)):\n",
    "        if statePrediction[i].isupper():\n",
    "            if island_start == None:\n",
    "                island_start = i\n",
    "        elif island_start != None:\n",
    "            print(island_start,\"to\",i)\n",
    "            island_start = None\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_000006.12:c32109338-32041153 Homo sapiens chromosome 6, GRCh38.p13 Primary Assembly\n",
      "0 to 159\n",
      "844 to 1042\n",
      "11936 to 13677\n",
      "20294 to 20521\n",
      "21150 to 22239\n",
      "23204 to 23516\n",
      "24656 to 24852\n",
      "27685 to 28063\n",
      "29992 to 30328\n",
      "38927 to 39840\n",
      "41369 to 41706\n",
      "44372 to 44497\n",
      "47633 to 47936\n",
      "50967 to 51306\n",
      "52474 to 52934\n",
      "53179 to 53480\n",
      "55177 to 55994\n",
      "56360 to 56661\n",
      "59062 to 59239\n",
      "59760 to 60100\n",
      "60617 to 61024\n",
      "61207 to 61528\n",
      "62136 to 63506\n",
      "63753 to 67624\n"
     ]
    }
   ],
   "source": [
    "main([\"TNXB_region.fasta\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">NC_000006.12:c32109338-32041153 Homo sapiens chromosome 6, GRCh38.p13 Primary Assembly\n",
    "0 to 159\n",
    "844 to 1042\n",
    "11936 to 13677\n",
    "20294 to 20521\n",
    "21150 to 22239\n",
    "23204 to 23516\n",
    "24656 to 24852\n",
    "27685 to 28063\n",
    "29992 to 30328\n",
    "38927 to 39840\n",
    "41369 to 41706\n",
    "44372 to 44497\n",
    "47633 to 47936\n",
    "50967 to 51306\n",
    "52474 to 52934\n",
    "53179 to 53480\n",
    "55177 to 55994\n",
    "56360 to 56661\n",
    "59062 to 59239\n",
    "59760 to 60100\n",
    "60617 to 61024\n",
    "61207 to 61528\n",
    "62136 to 63506\n",
    "63753 to 67624"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
